ledger/participant-integration-api/src/main/scala/platform/indexer/ha/PreemptableSequence.scala

```diff
@@ -0,0 +1,192 @@
+    * - the completion Future future of the PreemptableSequence will only finish after this Hanlde finishes,
+    *   and previously registered release functions all completed
+    * - KillSwitch events will be replayed to this handle
+    * - In case of abort/shutdown the PreemptableSequence's completion result will conform to the KillSwitch usage,
+    *   not to the completion of this handle (although it will wait for it naturally)
+    *
+    * @param handle The handle to delegate to
+    * @return the completion of the Handle
+    */
+  def merge(handle: Handle): Future[Unit]
+
+  /** The handle of the PreemptableSequence. This handle is available for sequence construction as well.
+    * @return the Handle
+    */
+  def handle: Handle
+}
+
+object PreemptableSequence {
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  /** @param executionContext this execution context will be used to:
+    *   - execute future transformations
+    *   - and encapsulate synchronous work in futures (this could be possibly blocking)
+    *   Because of the possible blocking nature a dedicated pool is recommended.
+    */
+  def apply(scheduler: Scheduler)(implicit
+      executionContext: ExecutionContext,
+      loggingContext: LoggingContext,
+  ): PreemptableSequence = { sequence =>
+    val resultCompleted = Promise[Unit]()
+    val killSwitchCaptor = new KillSwitchCaptor
+    val resultHandle = Handle(resultCompleted.future, killSwitchCaptor)
+    var releaseStack: List[() => Future[Unit]] = Nil
+
+    val helper: SequenceHelper = new SequenceHelper {
+      private def waitFor(delayMillis: Long): Future[Unit] =
+        goF(akka.pattern.after(FiniteDuration(delayMillis, "millis"), scheduler)(Future.unit))
+
+      override def registerRelease(release: => Unit): Unit = synchronized {
+        logger.info(s"Registered release function")
+        releaseStack = (() => Future(release)) :: releaseStack
+      }
+
+      override def goF[T](f: => Future[T]): Future[T] =
+        killSwitchCaptor.state match {
+          case _: KillSwitchCaptor.State.Used =>
+            // Failing Future here means we interrupt the Future sequencing.
+            // The failure itself is not important, since the returning Handle-s completion-future-s result is overridden in case KillSwitch was used.
+            logger.info(s"KillSwitch already used, interrupting sequence!")
+            Future.failed(new Exception("UsedKillSwitch"))
+
+          case _ =>
+            f
+        }
+
+      override def go[T](body: => T): Future[T] = goF[T](Future(body))
+
+      override def retry[T](waitMillisBetweenRetries: Long, maxAmountOfRetries: Long)(
+          body: => T
+      ): Future[T] =
+        go(body).transformWith {
+          // since we check countdown to 0, starting from negative means unlimited retries
+          case Failure(ex) if maxAmountOfRetries == 0 =>
+            logger.info(
```

Comment:
Shouldn't the level here be `warn`?

Reply:
yes, since we switched another place logging to warn which was about possibly non-transient error, I think here it is called for as well. changed